sklearn/utils/metadata_requests.py

```diff
@@ -0,0 +1,735 @@
+            )
+        if mapping == "one-to-one":
+            mapping = {method: method for method in METHODS}
+        other = metadata_request_factory(obj)
+        for destination, source in mapping.items():
+            my_method = getattr(self, destination)
+            other_method = getattr(other, source)
+            my_method.merge_method_request(
+                other_method,
+                overwrite=overwrite,
+                expected_metadata=expected_metadata,
+            )
+
+    def masked(self):
+        """Return a masked version of the requests.
+
+        A masked version is one which converts a ``{'prop': 'alias'}`` to
+        ``{'alias': True}``. This is desired in meta-estimators passing
+        requests to their parent estimators.
+        """
+        res = MetadataRequest()
+        for method in METHODS:
+            setattr(res, method, getattr(self, method).masked())
+        return res
+
+    def to_dict(self):
+        """Return dictionary representation of this object."""
+        output = dict()
+        for method in METHODS:
+            output[method] = getattr(self, method).requests
+        return output
+
+    def __repr__(self):
+        return self.to_dict()
+
+    def __str__(self):
+        return str(self.to_dict())
+
+
+def metadata_request_factory(obj=None):
+    """Get a MetadataRequest instance from the given object.
+
+    .. versionadded:: 1.1
+
+    Parameters
+    ----------
+    obj : object
+        If the object is already a MetadataRequest, return that.
+        If the object is an estimator, try to call `get_metadata_request` and get
+        an instance from that method.
+        If the object is a dict, create a MetadataRequest from that.
+
+    Returns
+    -------
+    metadata_requests : MetadataRequest
+        A ``MetadataRequest`` taken or created from the given object.
+    """
+    if obj is None:
+        return MetadataRequest()
+
+    if isinstance(obj, MetadataRequest):
+        return obj
+
+    if isinstance(obj, dict):
+        return MetadataRequest(obj)
+
+    try:
+        return obj.get_metadata_request(output="MetadataRequest")
+    except TypeError:
+        # The method exists, but doesn't accept `output`
+        return MetadataRequest(obj.get_metadata_request())
+    except AttributeError:
+        # The object doesn't have a `get_metadata_request` method.
+        return MetadataRequest()
+
+
+class MetadataRouter:
+    """Route the metadata to child objects.
+
+    .. versionadded:: 1.1
+    """
+
+    def __init__(self):
+        self.requests = MetadataRequest()
+
+    def add(self, *obj, mapping="one-to-one", overwrite=False, mask=False):
+        """Add a set of requests to the existing ones.
+
+        Parameters
+        ----------
+        *obj : objects
+            A set of objects from which the requests are extracted. Passed as
+            arguments to this method.
+
+        mapping : dict or str, default="one-to-one"
```

Comment:
are there cases requiring a one-to-many or many-to-many mapping?

Reply:
For instance, if a custom scorer is passed to a GridSearchCV, its requests are mapped to `fit` and `score`, so it'd be a one to many. But those cases are handled by a custom mapping dict. I'm not too attached to the `"one-to-one"` naming if that's what you're getting at. We could call it a bijection to confuse people :D

Reply:
you just expressed that in a dict :D I'm not sure what you mean by it can't be expressed.

Reply:
That dict is invalid! It has a duplicate key!

Reply:
:see_no_evil: I guess we could accept a list of tuples, but I find a list of tuples a very unintuitive construct, and wherever we have it in our API, I have to look it up again to see which member does what.

Reply:
On the other hand, we could add the list of tuples option later as an extra format. For now it seems to satisfy the needs.